pallets/liquidity-pools-gateway/src/tests.rs

```diff
@@ -530,36 +649,1896 @@ mod message_processor_impl {
+										message_submitted_times: 1,
+										expected_storage_entries: vec![
+											(
+												*ROUTER_HASH_1,
+												Some(InboundEntry::<Runtime>::Message {
+													domain_address: TEST_DOMAIN_ADDRESS,
+													message: Message::Simple,
+													expected_proof_count: 1,
+												}),
+											),
+											(*ROUTER_HASH_2, None),
+										],
+									},
+								),
+								(
+									vec![
+										(*ROUTER_HASH_1, Message::Simple),
+										(*ROUTER_HASH_2, Message::Proof(MESSAGE_PROOF)),
+										(*ROUTER_HASH_2, Message::Proof(MESSAGE_PROOF)),
+									],
+									ExpectedTestResult {
+										message_submitted_times: 1,
+										expected_storage_entries: vec![
+											(*ROUTER_HASH_1, None),
+											(
+												*ROUTER_HASH_2,
+												Some(InboundEntry::<Runtime>::Proof {
+													current_count: 1,
+												}),
+											),
+										],
+									},
+								),
+								(
+									vec![
+										(*ROUTER_HASH_2, Message::Proof(MESSAGE_PROOF)),
+										(*ROUTER_HASH_1, Message::Simple),
+										(*ROUTER_HASH_1, Message::Simple),
+									],
+									ExpectedTestResult {
+										message_submitted_times: 1,
+										expected_storage_entries: vec![
+											(
+												*ROUTER_HASH_1,
+												Some(InboundEntry::<Runtime>::Message {
+													domain_address: TEST_DOMAIN_ADDRESS,
+													message: Message::Simple,
+													expected_proof_count: 1,
+												}),
+											),
+											(*ROUTER_HASH_2, None),
+										],
+									},
+								),
+								(
+									vec![
+										(*ROUTER_HASH_2, Message::Proof(MESSAGE_PROOF)),
+										(*ROUTER_HASH_1, Message::Simple),
+										(*ROUTER_HASH_2, Message::Proof(MESSAGE_PROOF)),
+									],
+									ExpectedTestResult {
+										message_submitted_times: 1,
+										expected_storage_entries: vec![
+											(*ROUTER_HASH_1, None),
+											(
+												*ROUTER_HASH_2,
+												Some(InboundEntry::<Runtime>::Proof {
+													current_count: 1,
+												}),
+											),
+										],
+									},
+								),
+								(
+									vec![
+										(*ROUTER_HASH_2, Message::Proof(MESSAGE_PROOF)),
+										(*ROUTER_HASH_2, Message::Proof(MESSAGE_PROOF)),
+										(*ROUTER_HASH_1, Message::Simple),
+									],
+									ExpectedTestResult {
+										message_submitted_times: 1,
+										expected_storage_entries: vec![
+											(*ROUTER_HASH_1, None),
+											(
+												*ROUTER_HASH_2,
+												Some(InboundEntry::<Runtime>::Proof {
+													current_count: 1,
+												}),
+											),
+										],
+									},
+								),
+							]);
+
+						let suite = generate_test_suite(
+							vec![*ROUTER_HASH_1, *ROUTER_HASH_2],
+							TEST_DATA.clone(),
+							expected_results,
+							MESSAGE_COUNT,
+						);
+
+						run_inbound_message_test_suite(suite);
+					}
+				}
+
+				mod four_messages {
```

Comment:
If it works for 3, can we ensure it will always work for 4?

Reply:
In this case, we are testing with 2 routers, and the minimum number of messages needed for a successful message submission is 2. So, if it works with 2, it should work with more messages. 

Initially, I added tests for up to 4 messages to confirm that duplicate messages are processed accordingly. Still thinking about generating correct expected results so we can slim these tests down a bit.